sed/core.py

```diff
@@ -76,17 +97,48 @@ def load(self, data: pd.DataFrame) -> None:
         """
         self._dataframe = pd.DataFrame(data)
 
+    def gen_jittered_df(self, cols: Sequence[str] = None) -> None:
+        """Generate a a second dataframe, where the selected dataframe
+         columns have jitter applied
+
+        Args:
+            cols: the colums onto which to apply jitter. If omitted,
+            the comlums are taken from the config.
+
+        Returns:
+            None
+        """
+        if cols is None:
+            try:
+                cols = self._config["jitter_cols"]
+            except KeyError:
+                cols = self._dataframe.columns  # jitter all columns
+
+        self._dataframe_jittered = self._dataframe.map_partitions(
+            apply_jitter,
+            cols=cols,
+            cols_jittered=cols,
+        )
+
     def compute(
         self,
-        mode: str = "numba",
-        binDict: dict = None,
-        axes: list = None,
-        nbins: int = None,
-        ranges: list = None,
-        pbar: bool = True,
+        bins: Union[
+            int,
+            dict,
+            tuple,
+            List[int],
+            List[np.ndarray],
+            List[tuple],
+        ] = 100,
+        axes: Union[str, Sequence[str]] = None,
+        ranges: Sequence[Tuple[float, float]] = None,
+        histMode: str = "numba",
+        mode: str = "fast",
         jittered: bool = True,
-        ncores: int = N_CPU,
-        pbenv: str = "classic",
+        pbar: bool = True,
+        nCores: int = None,
```

Comment:
This is not a comment to this PR, but rather a general one. Do we want to stick to camelCase for function attributes and snake_case for variables? 
In any case, the parameter names used in the binning function should be the same. I'll set up an issue for this.

Reply:
Right, this is just copyPaste

Reply:
That's why I suggest to always write the code by yourself, even if you copy it from somewhere else ;)

Reply:
using all the cores freezes you computer. The best choice is N-1 cores. In some cases however, we have observed better performance using less cores, due to less overheads in the parallelization (I think..)

Reply:
Every "worker"/Core needs to reserve a copy of the target array in memory. For large binning sets and machines with many cores, you quickly run out of memory, if you don't limit the number of workers (e.g. 10 GB array* 140 cores = 1.4 TB RAM...). Also, you increase the overhead for recombination.

Reply:
> using all the cores freezes you computer. The best choice is N-1 cores. In some cases however, we have observed better performance using less cores, due to less overheads in the parallelization (I think..)

I assume from your words you expect the users are going to use this library locally. Only in that case, I would see a point for letting the number of cores as a parameter.

You're right regarding performance. There is an overhead due to parallelization but I thought the data we will be dealing with is big enough to compensate it.

Reply:
> Every "worker"/Core needs to reserve a copy of the target array in memory. For large binning sets and machines with many cores, you quickly run out of memory, if you don't limit the number of workers (e.g. 10 GB array* 140 cores = 1.4 TB RAM...). Also, you increase the overhead for recombination.

Why you need to copy data for every worker? That's not optimal at all.
Shared memory is used especifically for that. But you should know in advance the size of your data, which we know at the time of reading it.
Have a look at the following library we developed at EuXFEL (https://github.com/European-XFEL/pasha)

Reply:
Here, the issue is not about distributing the _source_ data over mutliple worker threads, this is done via dask, where every worker gets one dataframe partition. But each worker generates counts distributed over the _whole_ volume of the output array. If you want to use one common data volume for this, you need heavy synchronization between threads, with insane overhead, therefore one volume per thread, and later recombination. This works, is fast and reliable, but explodes if you make the number of workers too large. 20 is a good upper limit.

Reply:
Some benchmarks I did once:
3D binning:
![Binning 3D](https://user-images.githubusercontent.com/53396064/163780588-9fc4e983-aa16-4cd4-8ba2-f25951c50366.PNG)
![Speedup 3D](https://user-images.githubusercontent.com/53396064/163780629-b66856f4-9e76-497e-81aa-4cf6fb2671f0.PNG)
4D binning:
![Binning 4D](https://user-images.githubusercontent.com/53396064/163780611-108ec446-9bef-48b2-943d-5378117ee1aa.PNG)
![Speedup 4D](https://user-images.githubusercontent.com/53396064/163780647-33bbcfc3-84ff-494e-9277-8c5516329a3d.PNG)

Reply:
As said before, that's why the library mentioned before was built.
Dask is a black box where you have no control over the data/workers. I'm a bit biased against dask, because it was not easy to make it work in my use case. And it was not only me, my colleagues on the data analysis group do not use dask neither, because it's tricky to control it.
Simple parallezation and/or distributed computing is not so complicated in Python :)

Reply:
Using shared memory would be really nice. We should have a look into this (you told me already, but I didn't yet find a chance to..).

Reply:
We can discuss this off course, but I don't see a strong reason to change our running system.
We use dataframes not only for parallelization, but also for the delayed JIT loading of the data (which don't fit into memory), and the transparent and flexible manipulation of dataframe columns before computation without modification of the code. I am not sure how to reproduce these features otherwise, and even if we could, what the benefit of doing so would be.
In any case, this pull request is not the right place for such a discussion.

Reply:
First of all, as you correctly pointed out, this is the wrong place for such a discussion.
If we think we should look into this more, we should make an issue/discussion.

But as a quick response:
1. I am strongly against moving away from dask in this stage
2. We can test a binning aggregation method based on shared memory, but that would not remove the dask dependency as, like Laurenz pointed out, the delayed processing is fundamental for our workflow. It might be good to have both options maybe, but don't change a working system!